gnomad/utils/constraint.py

```diff
@@ -360,3 +412,279 @@ def collapse_strand(
+      and its value is a ArrayNumericExpression of observed variant counts for
+      specified populations. Default is {}.
+    :param weighted: Whether to generalize the model to weighted least squares using
+      'possible_variants'. Default is False.
+    :return: A dictionary of intercepts and slopes for plateau models of each
+      population. The key of the dictionary is a population name, and the value is a
+      dictionary (or a list of dictionaries if `pop_observed_variants_exprs` is
+      specified) mapping cpg BooleanExpression to a intercept and a slope.
+    """
+    # Build a plateau model using all the sites in the Table.
+    plateau_models_agg_expr = {
+        "total": hl.agg.group_by(
+            cpg_expr,
+            hl.agg.linreg(
+                observed_variants_expr / possible_variants_expr,
+                [1, mu_snp_expr],
+                weight=possible_variants_expr if weighted else None,
+            ).beta,
+        )
+    }
+    # Build plateau models using sites in population downsamplings if
+    # population is specified.
+    for pop, pop_observed_variants_expr in pop_observed_variants_exprs.items():
+        plateau_models_agg_expr[pop] = hl.agg.array_agg(
+            lambda pop_observed_variants: hl.agg.group_by(
+                cpg_expr,
+                hl.agg.linreg(
+                    pop_observed_variants / possible_variants_expr,
+                    [1, mu_snp_expr],
+                    weight=possible_variants_expr if weighted else None,
+                ).beta,
+            ),
+            pop_observed_variants_expr,
+        )
+    return plateau_models_agg_expr
+
+
+def build_coverage_model(
+    low_coverage_oe_expr: hl.expr.Float64Expression,
+    log_coverage_expr: hl.expr.Float64Expression,
+) -> hl.expr.StructExpression:
+    """
+    Build coverage model.
+
+    This function uses linear regression to build a model of log10(coverage) to correct proportion of expected variation at low coverage sites.
+
+    The x and y of the coverage model:
+    - x: log10('exome_coverage') at low coverage site
+    - y: sum('observed_variants')/ (`high_coverage_scale_factor` * sum('possible_variants' * 'mu_snp') at low coverage site
+    where `high_coverage_scale_factor` = sum('observed_variants') / sum('possible_variants' * 'mu_snp') at high coverage site
+
+    :param low_coverage_oe_expr: The Float64Expression of observed:expected ratio
+      for a given coverage level.
+    :param log_coverage_expr: The Float64Expression of log10 coverage.
+    :return: Tuple with intercept and slope of the model.
+    """
+    return hl.agg.linreg(low_coverage_oe_expr, [1, log_coverage_expr])
+
+
+def get_all_pop_lengths(
+    ht,
+    pops: Tuple[str],
+    prefix: str = "observed_",
+) -> List[Tuple[str, str]]:
+    """
+    Get the minimum length of observed variant counts array for each population downsamping.
+
+    The minimum array length will be used to determine the number of plateau models
```

Comment:
this isn't true anymore, right? since this function doesn't get called when building the plateau models